Apim 5001 - Category filter and query set as query params

jansobczynski opened the pull request:
## Issue

https://gravitee.atlassian.net/browse/APIM-5001

## Description

Category filter and query set as query params

## Additional context
Screenshot:
<img width="1182" alt="Screenshot 2024-05-16 at 1 45 07 PM" src="https://github.com/gravitee-io/gravitee-api-management/assets/144100648/579b725b-b374-4c4e-bba5-253b5a60aa0f">

<!-- Add any other context about the PR here -->
<!-- It can be links to other PRs or docs or drawing -->
<!-- Or reproduction steps in case of bug fix -->

<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-iinuwzqssa.chromatic.com)
<!-- Storybook placeholder end -->

jourdiw commented on gravitee-apim-portal-webui-next/src/components/api-filter/api-filter.component.ts:
```diff
@@ -66,6 +69,18 @@ export class ApiFilterComponent {
     }
   }
 
+  ngOnChanges(changes: SimpleChanges) {
```
Instead of using `ngOnChanges`, have `filterParam` as an InputSignal and add `effect(() -> {})` to the constructor.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/components/search-bar/search-bar.component.ts:
```diff
@@ -28,7 +28,10 @@ import { debounceTime, distinctUntilChanged } from 'rxjs';
-export class SearchBarComponent implements OnInit {
+export class SearchBarComponent implements OnInit, OnChanges {
+  @Input()
+  searchParam!: string;
```
Same here: make it an InputSignal to avoid having to implement OnChanges

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -76,11 +78,21 @@ export class CatalogComponent {
 
+  ngOnInit() {
+    this.route.queryParams.subscribe(params => {
+      this.selectedFilter = params['filter'];
```
You should be able to use just `@Input `+ `!` to access params.

jansobczynski replied:
We figured out that it's impossible to use it in child component which is not part of routing

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -92,11 +104,13 @@ export class CatalogComponent {
   public onFilterSelection(event: string) {
     this.selectedFilter = event;
     this.page$.next(1);
+    this.queryParam.updateParam({ filter: this.selectedFilter === 'all' ? '' : this.selectedFilter });
```
This should just be a simple navigate, right? You can have:

```typescript
this.router.navigate([''], {
      relativeTo: this.activatedRoute,
      queryParams: {
          filter: event,
          query: this.query
      },
    })
```

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -92,11 +104,13 @@ export class CatalogComponent {
   public onSearchResults(searchInput: string) {
     this.searchInput = searchInput;
     this.page$.next(1);
+    this.queryParam.updateParam({ query: this.searchInput });
```
Same here: not sure if it's really necessary to have a separate service to change the query param.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.html:
If you can get the inputs directly from the router in the api-filter and search-bar components, you can remove these changes since it will be inputted automatically by Angular.

jansobczynski committed:
feat(portal-next): category filter and query set as query params

feat(portal-next): category filter and query set as query params

feat(portal-next): category filter and query set as query params

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -92,19 +98,33 @@ export class CatalogComponent {
 
   public onFilterSelection(event: string) {
-    this.selectedFilter = event;
+    this.filter = event;
```
This will be updated after the navigate.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -92,19 +98,33 @@ export class CatalogComponent {
+    this.router.navigate([''], {
+      relativeTo: this.route,
+      queryParams: {
+        filter: this.filter === 'all' ? '' : this.filter,
```
Instead of `this.filter`, you can use the parameter `event`.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/app/catalog/catalog.component.ts:
```diff
@@ -92,19 +98,33 @@ export class CatalogComponent {
+        filter: this.filter,
+        query: this.query,
+      },
+    });
```
Same here: the navigate will update the query, no need to set it here.

jansobczynski replied:
Done

jourdiw commented on gravitee-apim-portal-webui-next/src/components/breadcrumb-navigation/breadcrumb-navigation.component.ts:
🗑️

jansobczynski replied:
I can't delete this component since there is breadcrumb logic inside html template

jansobczynski committed:
feat(portal-next): category filter and query set as query params
